### ISSUE TITLE
Multi-GPU Support

### ISSUE THREAD

**Issue by dkermany** [clEsperanto/pyclesperanto_prototype#195]
Are there any means of using multiple GPUs to boost available memory, or future plans of enabling support for this? I have access to several GPUs on my machine however I have not been able to find a means of utilizing more than one at a time. 

This is for running functions of very large 3D images where tiling produces unwanted results at the borders of the tiles.

**Comment by haesleinhuepf** [comment]
Hi @dkermany ,

I programmed this attempt towards multi-GPU support long time ago:
- https://github.com/clEsperanto/pyclesperanto_prototype/pull/129

It will be a lot of effort to finish this and that's why we're planning this for an upcoming major release (in a year or so).

> tiling produces unwanted results at the borders of the tiles.

I presume you are aware you can avoid this by processing image tiles with overlap as shown here: https://haesleinhuepf.github.io/BioImageAnalysisNotebooks/32_tiled_image_processing/tiling_images_with_overlap.html

Last but not least: would it be an option to process parts of your image(s) in separate scripts? E.g. you run one script on GPU1 and another script on GPU2?

**Comment by dkermany** [comment]
@haesleinhuepf thanks for such a rapid and helpful reply!

> It will be a lot of effort to finish this and that's why we're planning this for an upcoming major release (in a year or so).

Thank you for letting me know. This is a great library and I appreciate the effort and will keep an eye out

> I presume you are aware you can avoid this by processing image tiles with overlap as shown here:

I was NOT aware of this and how effective overlap can be! Thank you so much for reference!

> Last but not least: would it be an option to process parts of your image(s) in separate scripts? E.g. you run one script on GPU1 and another script on GPU2?

Would this not be identical to tiling but faster? Since processing is still done in the context of sub-images that need to be stitched together

**Comment by haesleinhuepf** [comment]
> Would this not be identical to tiling but faster? Since processing is still done in the context of sub-images that need to be stitched together

Yes, I was mostly suggesting this because it is technically feasible, fast (multi-GPU) tiling. 🌞
